pkg/validator/cluster.go

```diff
@@ -312,13 +312,61 @@ func validateGeneralClusterChanges(
 	valErrors []string,
 ) []string {
 
-	if cr.Spec.AppID != prevCr.Spec.AppID {
-		appModifiedMsg := fmt.Sprintf(
-			modifiedProperty,
-			"app",
-		)
-		valErrors = append(valErrors, appModifiedMsg)
+	crApp, err := catalog.GetApp(cr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
+	}
+
+	prevCrApp, err := catalog.GetApp(prevCr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
 	}
+
+	if crApp != nil && prevCrApp != nil {
+
+		rbKey := cr.Name + cr.Namespace
+		rbInfo := shared.ClusterRollbackInfoMap[rbKey]
+
+		if prevCrApp.Spec.DistroID != crApp.Spec.DistroID {
+			appModifiedMsg := fmt.Sprintf(
+				invalidDistroId,
+				crApp.Spec.DistroID,
+				prevCrApp.Spec.DistroID,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		if prevCrApp.Spec.Version == crApp.Spec.Version {
+			appModifiedMsg := fmt.Sprintf(
+				versionIsNotModified,
+				crApp.Spec.DistroID,
+				crApp.Spec.Version,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		if !prevCrApp.Spec.Upgradable {
+			appModifiedMsg := fmt.Sprintf(
+				appNotUpgradable,
+				prevCrApp.Spec.DistroID,
+				prevCrApp.Spec.Version,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		// Check, was the previous cluster configured succesfully
+		// If so, write its app identifiers as rollback info for the current one
+		if shared.ClusterIsReady(prevCr) {
```

Comment:
If the cluster isn't currently in the configured state I believe we should reject the change, as per the "strict" validation scheme from our discussion.

Reply:
Done. But I left an ability to apply the change for rollback application. A cluster may be in the `updating` state, and the proposed `AppID` should be equal to cluster `Status.UpgradeInfo.PrevApp` field and a `Status.UpgradeInfo.isRollingBack` field should be `false` (as we shouldn't interrupt the performed rollback process)